src/pallene/ir.lua

```diff
@@ -258,136 +265,117 @@ function ir.get_dsts(cmd)
+            if jmp_false then
+                depth_search(jmp_false.target)
+            end
+            order[#order + 1] = block_i
         end
+    end
 
-        -- Transform parent node
-        return f(cmd) or cmd
+    depth_search(1)
+    -- The actual block order is the reverse of what was calculated so far. We could have calculated
+    -- the right order from the start, but the existence of unreacheable blocks makes it harder than
+    -- usual (hence why we're doing it this way).
+    local reverse_order = {}
+    for i = #order, 1, -1 do
+        table.insert(reverse_order, order[i])
     end
-    return go(root_cmd)
-end
 
--- Remove some kinds of silly control flow
---   - Empty If
---   - if statements w/ constant condition
---   - Nop and Seq statements inside Seq
---   - Seq commands w/ no statements
---   - Seq commans w/ only one element
-function ir.clean(cmd)
-    local tag = cmd._tag
-    if tag == "ir.Cmd.Nop" then
-        return cmd
+    return reverse_order
+end
 
-    elseif tag == "ir.Cmd.Seq" then
-        local out = {}
-        for _, c in ipairs(cmd.cmds) do
-            c = ir.clean(c)
-            if c._tag == "ir.Cmd.Nop" then
-                -- skip
-            elseif c._tag == "ir.Cmd.Seq" then
-                for _, cc in ipairs(c.cmds) do
-                    table.insert(out, cc)
-                end
-            else
-                table.insert(out, c)
+-- Iterate over the cmds of basic blocks using a naive ordering.
+function ir.iter(block_list)
+    local function go()
+        for _,block in ipairs(block_list) do
+            for _, cmd in ipairs(block.cmds) do
+                coroutine.yield(cmd)
             end
         end
-        if     #out == 0 then
-            return ir.Cmd.Nop()
-        elseif #out == 1 then
-            return out[1]
-        else
-            return ir.Cmd.Seq(out)
-        end
-
-    elseif tag == "ir.Cmd.If" then
-        local v = cmd.src_condition
-        cmd.then_ = ir.clean(cmd.then_)
-        cmd.else_ = ir.clean(cmd.else_)
-        local t_empty = (cmd.then_._tag == "ir.Cmd.Nop")
-        local e_empty = (cmd.else_._tag == "ir.Cmd.Nop")
-
-        if t_empty and e_empty then
-            return ir.Cmd.Nop()
-        elseif v._tag == "ir.Value.Bool" and v.value == true then
-            return cmd.then_
-        elseif v._tag == "ir.Value.Bool" and v.value == false then
-            return cmd.else_
-        else
-            return cmd
-        end
-
-    elseif tag == "ir.Cmd.Loop" then
-        cmd.body = ir.clean(cmd.body)
-        return cmd
+    end
+    return coroutine.wrap(function() go() end)
+end
 
-    elseif tag == "ir.Cmd.For" then
-        cmd.body = ir.clean(cmd.body)
-        return cmd
+function ir.flatten_cmd(block_list)
+    local res = {}
+    for cmd in ir.iter(block_list) do
+        table.insert(res, cmd)
+    end
+    return res
+end
 
-    else
-        return cmd
+-- Remove jumps that are never taken
+function ir.clean(func)
```

Comment:
Future work: implement a better clean function such as the one described in Cooper&Torczon book. (please open an issue so we don't forget)

Reply:
#608